### PR TITLE
Curriculum-related additions: open elm, task encoder, manual curriculum

### PR DESCRIPTION
**Pull request by kywch**
Want to merge in open elm, task encoder, manual curriculum

* a skeleton for the end-to-end workflow in the `submission/submission_curriculum_track.py`
* a manually-curated training tasks in `curriculum/manual_curriculum.py`, with inputs from @dsctt
* an open-elm based task generator in the `elm_for_nmmo` folder (by @nishaanthkanna ), with a api-related tests in `tests/test_elm_for_nmmo.py`
* an LLM-based task encoder (by @nikhilpinnaparaju ), with a api-related tests in `tests/test_task_encoder.py`

TODO: replace these dummies in the `curriculum/train_helper.py` with the actual functions/classes
* task-conditional RL: `train_on_tasks`, `evaluate_on_tasks`, `load_agent_model`
* use leaning-progress based task sampler: `SyllabusTaskSampler`

### CODE REVIEW

File: curriculum/task_encoder.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+def extract_module_fn(module: ModuleType):
+  fn_dict = {}
+  for name, fn in module.__dict__.items():
+    if is_function_type(fn) and not name.startswith('_'):
+      fn_dict[name] = fn
+  return fn_dict
+
+######################################################################
+# Task embedding assumptions
+# CHECK ME: currently assuming each agent has ONLY ONE task assigned during training,
```
Comment (jsuarez5341) [inline curriculum/task_encoder.py:26]: Correct -- will actually be one per team... still thinking about that

File: curriculum/train_helper.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+"""this is a place holder for dummy functions"""
+# pylint: disable=all
+
+import random
+import numpy as np
+
+
+######################################################################
+# to be provided by Joseph
+#from cleanrl_ppo_lstm import train_on_tasks, evaluate_on_tasks
```
Comment (jsuarez5341) [inline curriculum/train_helper.py]: These will just be the train and evaluate methods of clean_pufferl once I add tasks as an arg

File: elm_for_nmmo/elm_helper.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  fn_dict = {}
+  for name, fn in module.__dict__.items():
+    if is_function_type(fn) and not name.startswith('_'):
+      fn_dict[name] = fn
+  return fn_dict
+
+PREBUILT_TASK_FN = extract_module_fn(nmmo.task.base_predicates)
+
+# used in OpenELMTaskGenerator: see self.config.env.impr = import_str["short_import"]
+import_str = {"short_import": """from predicates import TickGE,StayAlive,AllDead,EatFood,DrinkWater,CanSeeTile,CanSeeAgent,OccupyTile
```
Comment (jsuarez5341) [inline elm_for_nmmo/elm_helper.py:223]: What's all this?

File: elm_for_nmmo/nmmo_env.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    # else:
+    #   return None
+
+  @property
+  def fitness(self) -> Optional[float]:
+    return self._fitness
+
+
+# TODO: make the prompts, etc easy to edit
+class NMMOEnvironment(BaseEnvironment[NMMOTaskFn]):
```
Comment (jsuarez5341) [inline elm_for_nmmo/nmmo_env.py]: Why do we need to wrap the environment?

File: submission/submission_curriculum_track.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+EPOCHS_PER_ELM_UPDATE = 10
+
+# setting DEBUG=True bypass running open elm
+DEBUG = True
+
+def train_with_curriculum(use_elm=True):
+  # participant's custom task spec is provided in a separate file, custom_curriculum.py
+  if use_elm:
+    # NOTE: passing cc.task_spec just to get some tasks out
+    # CHECK ME: would it be realistic to generate the training task on the fly here?
```
Comment (jsuarez5341) [inline submission/submission_curriculum_track.py]: Possibly we can generate templates such as Equip <item> <level> and fill them in with scripts?

**Review by jsuarez5341 — APPROVED** [review]
Large PR, main thing from my first read through is eliminating env wrapper logic where possible. Is it not feasible to run ELM on nmmo.Env directly?

File: curriculum/task_encoder.py (hunk 1 above)
Comment (kywch) [reply curriculum/task_encoder.py:26]: a config/assert to ensure single task per agent was added to the env

finding a way to add embeddings of multiple tasks is a research question.

File: curriculum/train_helper.py (hunk 2 above)
Comment (kywch) [reply curriculum/train_helper.py]: Waiting for it to be finalized. The end-to-end script and env have been modified to work with a curriculum file, containing the whole training tasks and sampling training tasks on reset()

File: elm_for_nmmo/elm_helper.py (hunk 3 above)
Comment (kywch) [reply elm_for_nmmo/elm_helper.py:223]: These are info that are fed into LLM when running the ELM. We may or may not need to do so, and the elm-for-nmmo has a huge room for improvements.

File: elm_for_nmmo/nmmo_env.py (hunk 4 above)
Comment (kywch) [reply elm_for_nmmo/nmmo_env.py]: My current understanding is that open elm needs the env wrapper, like `BaseEnvironment`. We may or may not remove it, and I'll probably look into it later later.

File: submission/submission_curriculum_track.py (hunk 5 above)
Comment (kywch) [reply submission/submission_curriculum_track.py]: We may be able to use templates and/or prompts to "promote" diversity. What open elm does is basically prompting codegen model to generate code, and we can guide model by putting additional information in the prompt as code comments, like this: https://github.com/CarperAI/nmmo-baselines/blob/curriculum/elm_for_nmmo/nmmo_env.py#L161

**Comment by kywch** [comment]
The end-to-end training script has been modified to use the file-based curriculum, which the env `task-embed` branch now supports: https://github.com/CarperAI/nmmo-environment/pull/90

Also, the format of curriculum (i.e., task spec) has been refactored to use the `TaskSpec` class, instead of somewhat arbitrary task spec tuples, and it made the code cleaner. Please see the `TaskSpec` refactor PR on the env `class-spec` branch: https://github.com/CarperAI/nmmo-environment/pull/91

**Comment by kywch** [comment]
Resolved the conflicts with the 2.0 branch. Merging in.

**Pull request merged**